modules/torch/__init__.py

```diff
@@ -120,21 +151,76 @@ def hue_step(self, delta):
             self.led_h = 1 - HUE_STEP
         self.update_led()
 
+    def flash_stop(self):
+        if self.flash_mode == 0:
+            return
+        # Turn off flashing
+        self.flash_mode = 0
+        if self.timer is not None:
+            self.timer.cancel()
+            self.timer = None
+        # Turn off the LED
+        self.state = False
+        self.led_h = HUE_WHITE
+        self.led_v = 1.0
+        self.update_led()
+
+    def flash_change_mode(self):
+        self.flash_mode += 1
+        self.flash_state = 0
+        if self.flash_mode >= len(FLASH_PATTERNS):
+            self.flash_stop()
+            return
+        self.flash_set_state()
+
+    def flash_set_state(self):
+        new_state = FLASH_PATTERNS[self.flash_mode][1][self.flash_state]
+        if self.timer is not None:
+            self.timer.cancel()
+        self.timer = self.periodic(new_state[0], self.flash_led_cb)
+        self.state = True
+        self.led_h = new_state[1]
+        self.led_v = new_state[2]
+        self.update_led()
+
+    def flash_led_cb(self):
+        self.flash_state += 1
+        if self.flash_state >= len(FLASH_PATTERNS[self.flash_mode][1]):
+            self.flash_state = 0
+        self.flash_set_state()
+
     def on_start(self):
         super().on_start()
         self.state = False
         self.led_h = HUE_WHITE
         self.led_v = 1.0
         self.led = led
+        self.timer = None
+        self.flash_mode = 0   # Mode we're in.  0 means off.
+        self.flash_state = 0  # The part of the pattern we're in
+        self.is_active = True
 
         self.buttons.on_press(JOY_CENTRE, self.toggle_led)
         self.buttons.on_press(BUTTON_A, self.toggle_led)
         self.buttons.on_press(JOY_UP, self.brightness_up)
         self.buttons.on_press(JOY_DOWN, self.brightness_down)
         self.buttons.on_press(JOY_LEFT, lambda: self.hue_step(-HUE_STEP))
         self.buttons.on_press(JOY_RIGHT, lambda: self.hue_step(HUE_STEP))
+        self.buttons.on_press(BUTTON_B, self.flash_change_mode)
 
     def on_activate(self):
         super().on_activate()
+        self.is_active = True
```

Comment:
I think it'd be good to make is_active a property in `App` (backed by an `_is_active` member var, and have `App.on_activate()` and `App.on_deactivate()` set `_is_active`, or something like that)